### ISSUE TITLE
Blazor WASM with PWA and MapStaticAssets produce a wrong staticwebassets.endpoint.json file

### ISSUE THREAD

**Issue by DjeeBay** [dotnet/aspnetcore#59289]
When you have a Blazor WASM hosted app (created with a .Net 7 WASM template but upgraded to .Net 9) with PWA enabled, you have an integrity check failed issue when you try to use the new feature Asset["BlazorApp.styles.css"] where BlazorApp.styles.css is the generated css file (concatenation of component scoped css files).

If you check the **service-worker-assets.js** you'll see something like :
`{
      "hash": "sha256-PH+rhSKHi1BIxSZmJmXJLXV+HhUwVkPWEsu4GSnnegw=",
      "url": "BlazorApp.styles.css"
    },`

But in the **staticwebassets.endpoints.json** you don't have any corresponding route (that indicates the final name with the version number) for this file.
For instance, app.css (the file that is by default in wwwroot of the Client project) you will have something like :
`"Route": "css/app.7wkt1ejxmu.css",
      "AssetFile": "css/app.css",`

The consequence is that locally everything works, but a published/prod app will try to check the integrity of a file that don't really exists because the path to BlazorApp.styles.css is a 404 :
`Failed to find a valid digest in the 'integrity' attribute for resource 'https://mydomain.com/BlazorApp.Client.styles.css' with computed SHA-256 integrity '47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU='. The resource has been blocked.`

The only difference between app.css (or any other css in wwwroot) is that BlazorApp.styles.css is a generated file that just do an @import of the Client css file that is also generated :
`@import 'BlazorApp.Client.kayaoleqja.bundle.scp.css';`

The workaround was to exclude the 'BlazorApp.styles.css' file in the service-worker.published.js file :
`const offlineAssetsExclude = [ /^service-worker\.js$/, /^BlazorApp\.Client\.styles\.css$/ ];`

### .NET Version

9.0.100

**Comment by javiercn** [comment]
@DjeeBay thanks for contacting us.

Have you cleaned up the `bin` and `obj` folders after you did the TFM upgrade?

**Comment by DjeeBay** [comment]
@javiercn yes, it's deployed with docker so it's always like a first build.

**Comment by javiercn** [comment]
@DjeeBay can you share a repro project?

Also, is there a reason you started with a 7.0 template?

**Comment by MR-MAP** [comment]
@DjeeBay  Find Any solutions ?

**Comment by DjeeBay** [comment]
@MR-MAP I've found a workaround : you can exclude files from caching in the service-worker.published.js.

`const offlineAssetsExclude = [ /^service-worker\.js$/, /^BlazorApp\.Client\.styles\.css$/ ];`

**Comment by kibblewhite** [comment]
I've come across somewhat the same issue and not been able to resolve it.

**Comment by Xenogix** [comment]
We're experiencing the same issue in our Blazor WASM hosted PWA. The problem with excluding the bundled CSS file is that it breaks offline support for the PWA.

I came across [this related issue](https://github.com/dotnet/aspnetcore/issues/50856), which seems very similar.

**Comment by javiercn** [comment]
@DjeeBay We have investigated the issue and can explain what is happening.

## What we observed

In your project created from the Blazor Web template, the `StaticWebAssetProjectMode` property is being explicitly overridden from its default value of `Root` (which is the standard for projects using the BlazorWebAssembly SDK) to `Default`. This override causes the generated scoped CSS file `BlazorAppPwaAssetsIntegrityFails.Client.styles.css` to appear in the service worker manifest but not in the static web assets endpoints file, resulting in 404 errors and integrity validation failures when the service worker attempts to cache this asset.

## Why this happens

The Blazor Web template explicitly sets `StaticWebAssetProjectMode` to `Default` to integrate the WebAssembly client project as a referenced component within the hosted application structure. However, the BlazorWebAssembly SDK naturally defaults to `Root` mode because WebAssembly projects are designed to be self-contained applications with their own asset management.

When the template overrides this to `Default`, it creates an impedance mismatch: the service worker (which operates at the client project level) still discovers and attempts to cache all client assets including the generated CSS file, but the static web assets pipeline (now in `Default` mode) doesn't expose this generated file as a routable endpoint. The SDK's PWA features expect the `Root` behavior where all client assets are properly surfaced.

## Recommended approaches or alternatives

**Option 1: Remove the StaticWebAssetProjectMode override** (Recommended)

Since the BlazorWebAssembly SDK already defaults to `Root`, simply remove the explicit property setting to restore the correct behavior:

````xml
<Project Sdk="Microsoft.NET.Sdk.BlazorWebAssembly">

  <PropertyGroup>
    <TargetFramework>net10.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <ServiceWorkerAssetsManifest>service-worker-assets.js</ServiceWorkerAssetsManifest>
    <NoDefaultLaunchSettingsFile>true</NoDefaultLaunchSettingsFile>
    <!-- Remove this line entirely to use the SDK default of Root -->
  </PropertyGroup>

  <!-- ...existing code... -->
</Project>
````

This allows the client project to function as the self-contained WebAssembly application it's designed to be, with proper asset endpoint mapping for all generated files.

**Option 2: Move service worker to the Server project**

If the template's override was intentional for specific hosting scenarios, consider relocating the service worker configuration to the server project. This approach would create a unified manifest that properly handles assets from both projects under the server's asset management scope.

The mismatch between the Blazor Web template's configuration and the BlazorWebAssembly SDK's PWA expectations is being tracked. Please refer to [issue #48935](https://github.com/dotnet/aspnetcore/issues/48935) for updates on adding proper PWA support to the Blazor Web template that would eliminate these configuration conflicts.
